vendor/github.com/swiftstack/sortedmap/btree.go

```diff
@@ -1232,6 +1236,31 @@ func (tree *btreeTreeStruct) insertHere(insertNode *btreeNodeStruct, key Key, va
 	return
 }
 
+func (tree *btreeTreeStruct) AssertNodeEmpty(node *btreeNodeStruct) {
```

Comment:
Cool function - but we might want to return err to the caller in this case.
That is the pattern adopted elsewhere i.e. non-nil err always means corruption.
Where "errors" are "occasionally ok", I've also returned an ok bool e.g. when the caller does a Get on a non-existent key.

Reply:
I actually think that's a bad pattern!  If the tree is now corrupt _due to an internal logic error in the code_ then returning an error to the caller seems pointless.  What is the caller supposed to do???  The tree should not be accessed again.  Certainly it will be difficult to debug the situation with all the context lost (code at the Proxyfs layer looking at an error return).

Note that it is possible to catch a panic so a caller could catch this type of error if they wanted to (and log a useful error message).  I suppose a caller that wanted to recover could drop the tree on the ground and try starting over with a new tree ... though i don't think Proxyfs would be able to do that.

I do draw a distinction between internal logic errors and errors caused for external reasons, such as a bad value passed by the caller or being unable to read a node from disk (which _will happen_ and is not the fault of the the sortedmap code).  But, as near as i can think of failing to read a node from disk will not result in a corrupt tree, it just means that some nodes are inaccessible (either temporarily or permanently).  I think continuing to use the rest of the tree should be fine.  Having part of the tree inaccessible is a different situation then, say, inserting a new key and getting half way up the tree splitting nodes only to discover that some parent node can no longer perform Len().

This would require some careful thought about nodes that are readable from disk but turn out to be corrupt.  I don't think we want to panic in that situation.

I gave some thought to converting all those error cases that should never happen into panic()s, but decided to go with the smaller change and discuss this issue later (like now!).

Reply:
You and I are on the same page w.r.t. logic errors... but others have a different opinion.
I'd rather make a decision on the pattern and implement it consistently. I'm merely saying that your change makes this pattern inconsistent.

Let's chat about it tomorrow - I think it's a good discussion to have.